app/javascript/src/LoginPage/loginForms/SignupForm.jsx

```diff
@@ -0,0 +1,170 @@
+
+type Props = {
+  path: string
+}
+
+type State = {
+  username: string,
+  emailAddress: string,
+  password: string,
+  passwordConfirmation: string,
+  isValidUsername: ?boolean,
+  isValidEmailAddress: ?boolean,
+  isValidPassword: ?boolean,
+  isValidPasswordConfirmation: ?boolean
+}
+type HandlersProps = {
+  handleTextInputUsername: (username: string) => void,
+  handleTextInputEmail: (emailAddress: string) => void,
+  handleTextInputPassword: (password: string) => void,
+  handleTextInputPasswordConfirmation: (passwordConfirmation: string) => void
+}
+
+const FormGroups = ({state, handlers}: {state: State, handlers: HandlersProps}) => {
+  const {username, isValidUsername, emailAddress, isValidEmailAddress, password, isValidPassword, passwordConfirmation, isValidPasswordConfirmation} = state
+  const {handleTextInputUsername, handleTextInputEmail, handleTextInputPassword, handleTextInputPasswordConfirmation} = handlers
+
+  const usernameInputProps = {
+    value: username,
+    onChange: handleTextInputUsername,
+    autoFocus: 'autoFocus',
+    inputIsValid: isValidUsername
+  }
+  const emailInputProps = {
+    value: emailAddress,
+    onChange: handleTextInputEmail,
+    autoFocus: 'autoFocus',
+    inputIsValid: isValidEmailAddress
+  }
+  const passwordInputProps = {
+    value: password,
+    onChange: handleTextInputPassword,
+    ariaInvalid: 'false',
+    inputIsValid: isValidPassword
+  }
+  const passwordConfirmationInputProps = {
+    value: passwordConfirmation,
+    onChange: handleTextInputPasswordConfirmation,
+    ariaInvalid: 'false',
+    inputIsValid: isValidPasswordConfirmation
+  }
+
+  return (
+    <React.Fragment>
+      <FormGroup
+        type='user[username]'
+        labelIsValid={isValidUsername}
+        inputProps={usernameInputProps}
+      />
+      <FormGroup
+        type='user[email]'
+        labelIsValid={isValidEmailAddress}
+        inputProps={emailInputProps}
+      />
+      <FormGroup
+        type='user[password]'
+        labelIsValid={isValidPassword}
+        inputProps={passwordInputProps}
+      />
+      <FormGroup
+        type='user[password_confirmation]'
+        labelIsValid={isValidPasswordConfirmation}
+        inputProps={passwordConfirmationInputProps}
+      />
+    </React.Fragment>
+  )
+}
+
+class SignupForm extends Component<Props, State> {
+  constructor (props: Props) {
+    super(props)
+    this.state = {
+      username: '',
+      emailAddress: '',
+      password: '',
+      passwordConfirmation: '',
+      isValidUsername: undefined,
+      isValidEmailAddress: undefined,
+      isValidPassword: undefined,
+      isValidPasswordConfirmation: undefined
+    }
+  }
+
+  setIsValidUsername = () => this.setState({ isValidUsername: this.state.username !== '' })
+
+  handleTextInputUsername = (username: string) => this.setState({ username }, this.setIsValidUsername)
+
+  setIsValidEmail = () => this.setState({ isValidEmailAddress: this.state.emailAddress !== '' })
```

Comment:
I don't think this kind of validation is really useful this way.... Isn't PF4 providing a way to validate email? This applies to the rest of the validation in this form. You might want to use HTML5 validation instead.

Reply:
No, PF4 is not providing any kind of validation.
We can add HTML validation easily

Reply:
Then all of those functions can be reduce to one: `required`